Improve conflict handling of lifecycle mappings #549

laeubi opened the pull request:
This is a first step improving the following things

- record what is causing the conflict
- add the exception to the maven problem
- show problem markers at the correct location in the pom
- store the conflicting bundles at the marker

laeubi commented:
As this already has visual and useful enhancement for the user I'd like to merge this as a first step.

The next step then should be to have a quick-fix to actually ignore a certain conflicting mapping.

HannesWell commented on org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/lifecyclemapping/DuplicatePluginExecutionMetadataException.java:
```diff
@@ -0,0 +1,42 @@
+ * DuplicatePluginExecutionMetadataException
+ *
+ */
+public class DuplicatePluginExecutionMetadataException extends DuplicateMappingException{
```
Here Eclipse complains about a missing "static final serialVersionUID field of type long" as well.

laeubi replied:
Done

HannesWell commented on org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/lifecyclemapping/DuplicateMappingSourceProblem.java:
```diff
@@ -0,0 +1,82 @@
+      Version version = bundle.getVersion();
+      String v = version.getMajor() + "." + version.getMinor() + "." + version.getMicro();
+      marker.setAttribute(prefix + "bundleName",
+          Objects.requireNonNullElse(bundle.getHeaders().get(Constants.BUNDLE_NAME), bundle.getSymbolicName()));
```
Why do you check the headers first? Can the bundle.getSymbolicName() be `null`?

laeubi replied:
I'd just like to get the "name" of the bundle as a human readable name, so we later can use this in a quickfix and falling back to the symbolic name if non is given. The symbolic name should always be there.

HannesWell reviewed (commented):
This is indeed already an improvement and the code looks good in general, I just added few minor remarks.
Do you have a simple example that can be used as a test-case?

Some places like `FailedMappingMetadataSource` could be polished a bit, IMHO the (non-) javadoc is not very useful. Also the constructors of MavenProblemInfo could be further streamlined. I can add corresponding suggestions here or add a follow up PR. As you prefer.

laeubi commented:
> I can add corresponding suggestions here or add a follow up PR.

Follow Up is a bit better, you might also want to take a look at the eclipse settings here we ship they generate javadoc with copyright info of sonatype in 2010 whenever one creates a new file ...

laeubi commented:
> Do you have a simple example that can be used as a test-case?

It is really a bit tricky to actually test this, I have done the following:

1. Import the https://github.com/tesla/m2eclipse-tycho into my workspace
2. Add a stripped down version of [the sling tooling ](https://github.com/apache/sling-ide-tooling/tree/93bcb69fe7cfad6b68d5caf838a14d3bdf887b92/eclipse/eclipse-m2e-ui) to the workspace where I deleted all but `BndMavenPluginProjectConfigurator` and `MavenBundlePluginProjectConfigurator` and replaced them with a dummy implementation doing simply nothing.
3. Started the dev-eclipse and created a bnd and felix bundle project [test-projects.zip](https://github.com/eclipse-m2e/m2e-core/files/8012721/test-projects.zip)
.

laeubi commented on org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/lifecyclemapping/DuplicateLifecycleMappingMetadataException.java:
```diff
@@ -0,0 +1,43 @@
+ * DuplicateLifecycleMappingMetadataException
+ *
+ */
+public class DuplicateLifecycleMappingMetadataException extends DuplicateMappingException {
```
Done.

laeubi commented:
@HannesWell rebased and updated, if the build is fine and there are no further comments I'd like to merge this to start with the quick-fixes.

HannesWell commented:
> > I can add corresponding suggestions here or add a follow up PR.
> 
> Follow Up is a bit better, you might also want to take a look at the eclipse settings here we ship they generate javadoc with copyright info of sonatype in 2010 whenever one creates a new file ...

Alright 👍🏽 
That should definitely be fixed. I already thought multiple times that we should unify the settings (especially regarding formatting) in the entire m2e-project.

HannesWell reviewed (approved):
Thanks for sharing the steps to reproduce. Looks indeed a bit cumbersome, so I will skip this part for now.

> @HannesWell rebased and updated, if the build is fine and there are no further comments I'd like to merge this to start with the quick-fixes.

No further comments from my side. Thank you!
